### PR TITLE
fix unicorn service `upgrade` command

### PR DESCRIPTION
**Pull request by wjordan**
don't quit old process until new one is fully initialized.

a few weeks ago @mehalshah brought to my attention that adhoc instances would return 503 errors for about a minute after running `sudo service dashboard upgrade` before starting to serve traffic again. I found this odd because I had assumed the `upgrade` target would cut over traffic to the newly-created Unicorn process only when it was fully preloaded and ready to serve traffic. However, it turns out that the example script we were using simply cut over traffic immediately (well, after `sleep 2`), leaving a minute or so of downtime while the script preloading initializer took place on the new process. This was the cause not only of adhoc-restart issues, but also (I suspect) of intermittent downtime during rolling-restart deploys with low numbers (3 or less) of instances currently active.

The solution is not to kill the old Unicorn process from the shell script, but instead to kill it within a `before_fork` Ruby block that gets executed only after the application is fully preloaded.

While I was messing around with this target, I've also pulled in the newest upstream [changes](http://bogomips.org/unicorn.git/commit/examples/init.sh?id=6a65054ce2763d51aca5da525656bcb1f08d61a5) to this target committed this June:

> Rework the "upgrade" target to only read the PID files once to
> avoid misreading the wrong PID files in the middle of the
> upgrade.

### CODE REVIEW

File: pegasus/config/unicorn.rb [hunk 1]
```diff
@@ -7,3 +7,15 @@
...
+  # Quit the old unicorn process
+  old_pid = "#{server.config[:pid]}.oldbin"
+  if File.exist?(old_pid) && server.pid != old_pid
+    begin
+      Process.kill("QUIT", File.read(old_pid).to_i)
+    rescue Errno::ENOENT, Errno::ESRCH
+      # someone else did our job for us
+    end
+  end
+end
```
Comment (ashercodeorg) [inline pegasus/config/unicorn.rb]: Is it worth refactoring this into a method and calling the method from both dashboard and pegasus `unicorn.rb`? Probably not? Is it worth adding a comment that this code is duplicated (to both places)? Maybe?

**Comment by ewjordan** [comment]
LGTM in the abstract, though as with anything shell-related it's real easy to miss stuff - I don't entirely understand the reasoning behind a lot of the unicorn.sh changes coming from upstream, though I trust they've probably been vetted pretty thoroughly there.

Have you tested this out much? I'm wondering if it might be worth pushing this out in its own DTP, so as not to put a normal DTP with feature work at risk.

File: pegasus/config/unicorn.rb (hunk 1 above)
Comment (wjordan) [reply pegasus/config/unicorn.rb]: after a bit of experimentation, managed to refactor into a `Proc` shared between both config files.

**Comment by wjordan** [comment]
thanks for taking a look guys! @ewjordan I've tested this manually on adhoc instances.

Good thought on separating out the merge, however since this feature requires a Chef-cookbook change it's a bit difficult (would require blocking the whole staging-test pipeline rather than a hotfix and back-merge) to include this in a standalone deploy, and I don't think this item is important/risky enough to block the whole pipeline. So instead I'll just closely monitor the deploy when it goes out (it will get run in staging+test as well before it hits production) and be prepared to manually recover should any issues arise.

**Comment by wjordan** [comment]
Actually, it might be possible to hotfix+back-merge after this gets merged into staging, once the cookbook-update commit is done. I'll try that approach if it looks possible.

**Pull request merged**